### PR TITLE
[Thread] add thread pskc to network provisioning

### PR DESCRIPTION
**Pull request by gjc13**
This change allow users to set PSKc in network provisioning.
PSKc is used to authenticate thread commissioner. The unimplemented
part of PSKc configuration is completed.

### CODE REVIEW

File: src/device-manager/java/WeaveDeviceManager-JNI.cpp [hunk 1]
```diff
@@ -3051,12 +3051,15 @@ WEAVE_ERROR J2N_NetworkInfo(JNIEnv *env, jobject inNetworkInfo, NetworkInfo& out
 
     err = J2N_ByteArrayFieldVal(env, inNetworkInfo, "ThreadExtendedPANId", outNetworkInfo.ThreadExtendedPANId, len);
```
Comment (robszewczyk) [inline src/device-manager/java/WeaveDeviceManager-JNI.cpp]: Missing from the PR:  modifications to `N2J_NetworkInfo` to account for PSKc, changes to `NetworkInfo.java` to account for a new field, changes to the constructor for above.

File: src/device-manager/java/WeaveDeviceManager-JNI.cpp [hunk 2]
```diff
@@ -3128,6 +3138,7 @@ WEAVE_ERROR N2J_NetworkInfo(JNIEnv *env, const NetworkInfo& inNetworkInfo, jobje
                                                  threadNetName,
                                                  threadExtPANId,
                                                  threadKey,
+                                                 threadPSKc,
```
Comment (robszewczyk) [inline src/device-manager/java/WeaveDeviceManager-JNI.cpp]: `Make` method signature needs to be modified in line 3126 above

File: src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java [hunk 3]
```diff
@@ -45,6 +45,7 @@ public NetworkInfo()
         ThreadNetworkName = null;
         ThreadExtendedPANId = null;
         ThreadNetworkKey = null;
+        ThreadPSKc = null;
```
Comment (robszewczyk) [inline src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java]: A bunch of other code in `Make` and `MakeThread` below will need to be modified to account for a new parameter.

Comment (robszewczyk) [reply src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java]: The work items listed here remain unaddressed.

**Comment by gjc13** [comment]
The request changes have all been updated.

**Comment by gjc13** [comment]
@robszewczyk @gerickson @jaylogue 
Friendly reminder :)

File: src/device-manager/python/weave-device-mgr.py [hunk 4]
```diff
@@ -1284,6 +1289,7 @@ def do_createthreadnetwork(self, line):
         optParser = OptionParser(usage=optparse.SUPPRESS_USAGE, option_class=ExtendedOption)
         optParser.add_option("-n", "--name", action="store", dest="threadNetworkName", type="string")
         optParser.add_option("-k", "--key", action="store", dest="threadNetworkKey", type="string")
+        optParser.add_option("-s", "--pskc", action="store", dest="threadPskc", type="string")
```
Comment (bukepo) [inline src/device-manager/python/weave-device-mgr.py]: use `-C` to keep consistent with *otbr-commissioner*? And unify the variable name to be `threadPSKc`?

File: src/lib/profiles/network-provisioning/NetworkInfo.h [hunk 5]
```diff
@@ -73,17 +73,28 @@ class NL_DLL_EXPORT NetworkInfo
...
                                      * owned by the class.  Destroyed on any condition that calls `Clear()` on
                                      * the object. */
     uint8_t *ThreadNetworkKey;      /**< The Thread master network key , or NULL if not specified. It is a dynamically
-                                     * allocated array of arbitrary octets, owned by the class, with length specified
-                                     * by `ThreadNetworkKeyLen`.  Destroyed on any condition that calls `Clear()` on
+                                     * allocated array of arbitrary octets, owned by the class
+                                     * Destroyed on any condition that calls `Clear()` on
+                                     * the object. */
+    uint8_t *ThreadPSKc;            /**< Thread pre-shared key for commissioner, or NULL if not specified. It is a
+                                     * dynamically * allocated array of arbitrary octets, owned by the class
```
Comment (bukepo) [inline src/lib/profiles/network-provisioning/NetworkInfo.h]: what's the `*` after `dynamically` for?

File: src/device-manager/python/openweave/WeaveDeviceMgr.py [hunk 6]
```diff
@@ -354,7 +359,7 @@ class _NetworkInfoStruct(Structure):
         ('ThreadNetworkName', c_char_p),    # The name of the Thread network.
         ('ThreadExtendedPANId', c_void_p),  # The Thread extended PAN id (8 bytes).
         ('ThreadNetworkKey', c_void_p),     # The Thread master network key.
-        ('ThreadNetworkKeyLen', c_uint32),  # The length in bytes of the Thread master network key.
+        ('ThreadPSKc', c_void_p),           # The Thread preshared key for commissioner
```
Comment (bukepo) [inline src/device-manager/python/openweave/WeaveDeviceMgr.py]: preshared -> pre-shared

File: src/device-manager/python/weave-device-mgr.py (hunk 4 above)
Comment (gjc13) [reply src/device-manager/python/weave-device-mgr.py]: I'm not sure whether we tend to use capital letters in weave-device-manger.

Comment (bukepo) [reply src/device-manager/python/weave-device-mgr.py]: @robszewczyk Thoughts?

File: src/device-manager/python/weave-device-mgr.py [hunk 7]
```diff
@@ -138,6 +142,42 @@ def DecodeHexIntOption(option, opt, value):
...
+def aes_cmac_prf_128(key, message):
+    if len(key) != 16:
+        key = aes_cmac(('00' * 16).decode('hex'), key)
+    return aes_cmac(key, message)
+
+
+def compute_pskc(passphrase, network_name, network_xpanid):
+    print(passphrase)
+    print(network_name)
+    print(network_xpanid)
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: Remove prints.

File: src/device-manager/python/weave-device-mgr.py [hunk 8]
```diff
@@ -1216,27 +1260,45 @@ def do_addthreadnetwork(self, line):
...
             try:
-                networkInfo.SetField(name, val)
+                if name != 'passphrase':
+                    networkInfo.SetField(name, val)
             except Exception, ex:
                 print str(ex)
                 return
 
-        if networkInfo.ThreadPANId  != None:
+        if passphrase:
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: != None

File: src/device-manager/python/weave-device-mgr.py [hunk 9]
```diff
@@ -1216,27 +1260,45 @@ def do_addthreadnetwork(self, line):
...
+            if networkInfo.ThreadPSKc:
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: != None

File: src/device-manager/python/weave-device-mgr.py [hunk 10]
```diff
@@ -1216,27 +1260,45 @@ def do_addthreadnetwork(self, line):
...
+                print('PSKc already specified')
+                return
+            if not networkInfo.ThreadNetworkName:
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: == None

File: src/device-manager/python/weave-device-mgr.py [hunk 11]
```diff
@@ -1216,27 +1260,45 @@ def do_addthreadnetwork(self, line):
...
+                print('Thread network name not specified, cannot compute pskc')
+            if not networkInfo.ThreadExtendedPANId:
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: == None

File: src/device-manager/python/weave-device-mgr.py [hunk 12]
```diff
@@ -1216,27 +1260,45 @@ def do_addthreadnetwork(self, line):
...
                 print str(ex)
                 return
 
-        if networkInfo.ThreadPANId  != None:
+        if passphrase:
+            if networkInfo.ThreadPSKc:
+                print('PSKc already specified')
+                return
+            if not networkInfo.ThreadNetworkName:
+                print('Thread network name not specified, cannot compute pskc')
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: return?

File: src/device-manager/python/weave-device-mgr.py [hunk 13]
```diff
@@ -1216,27 +1260,45 @@ def do_addthreadnetwork(self, line):
...
+            if not networkInfo.ThreadExtendedPANId:
+                print('Thread xpanid not specified, cannot compute pskc')
```
Comment (jaylogue) [inline src/device-manager/python/weave-device-mgr.py]: return?

File: src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java (hunk 3 above)
Comment (robszewczyk) [reply src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java]: I stand by my previous comment.  I would expect a few more changes in NetworkInfo.java

File: src/device-manager/cocoa/NLNetworkInfo.mm [hunk 14]
```diff
@@ -163,9 +162,8 @@ - (NetworkInfo)toNetworkInfo
     }
 
     if (_ThreadNetworkKey) {
-        networkInfo.ThreadNetworkKeyLen = [_ThreadNetworkKey length];
-        networkInfo.ThreadNetworkKey = (uint8_t *) malloc(networkInfo.ThreadNetworkKeyLen);
-        memcpy(networkInfo.ThreadNetworkKey, [_ThreadNetworkKey bytes], networkInfo.ThreadNetworkKeyLen);
+        networkInfo.ThreadNetworkKey = (uint8_t *) malloc(NetworkInfo::kThreadNetworkKeyLength);
+        memcpy(networkInfo.ThreadNetworkKey, [_ThreadNetworkKey bytes], NetworkInfo::kThreadNetworkKeyLength);
```
Comment (robszewczyk) [inline src/device-manager/cocoa/NLNetworkInfo.mm]: Thread PSKc seems not to be exported into the Objective C land.

Comment (gjc13) [reply src/device-manager/cocoa/NLNetworkInfo.mm]: Updated

File: src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java (hunk 3 above)
Comment (robszewczyk) [reply src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java]: Please apply the following patch:

From 8ece3473237630f331257f022be6b41379201697 Mon Sep 17 00:00:00 2001
From: Robert Szewczyk <rszewczyk@nestlabs.com>
Date: Fri, 2 Aug 2019 13:54:06 -0700
Subject: [PATCH] Outstanding changes to fully flesh out PSKc support.

---
 .../java/src/nl/Weave/DeviceManager/NetworkInfo.java         | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java b/src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java
index 71de17a6..e157182e 100644
--- a/src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java
+++ b/src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java
@@ -151,7 +151,7 @@ public class NetworkInfo
     public static NetworkInfo Make(int networkType, long networkId,
                                    String wifiSSID, int wifiMode, int wifiRole, int wifiSecurityType, byte[] wifiKey,
                                    String threadNetworkName, byte[] threadExtendedPANId, byte[] threadNetworkKey,
-                                   short wirelessSignalStrength, int threadPANId, int threadChannel)
+                                   byte[] threadPSKc, short wirelessSignalStrength, int threadPANId, int threadChannel)
     {
         NetworkInfo netInfo = new NetworkInfo();
         netInfo.NetworkType = nl.Weave.DeviceManager.NetworkType.fromVal(networkType);
@@ -164,6 +164,7 @@ public class NetworkInfo
         netInfo.ThreadNetworkName = threadNetworkName;
         netInfo.ThreadExtendedPANId = threadExtendedPANId;
         netInfo.ThreadNetworkKey = threadNetworkKey;
+        netInfo.ThreadPSKc = threadPSKc;
         netInfo.ThreadPANId = threadPANId;
         netInfo.ThreadChannel = threadChannel;
         netInfo.WirelessSignalStrength = wirelessSignalStrength;
@@ -176,7 +177,7 @@ public class NetworkInfo
                                    short wirelessSignalStrength)
     {
         return Make(networkType, networkId, wifiSSID, wifiMode, wifiRole, wifiSecurityType, wifiKey, threadNetworkName,
-                    threadExtendedPANId, threadNetworkKey, wirelessSignalStrength, THREAD_PANID_NOTSPECIFIED,
+                    threadExtendedPANId, threadNetworkKey, null, wirelessSignalStrength, THREAD_PANID_NOTSPECIFIED,
                     THREAD_CHANNEL_NOTSPECIFIED);
     }
 }
-- 
2.22.0.770.g0f2c4a37fd-goog

Comment (robszewczyk) [reply src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java]: The above patch should preferably be squashed into one of your commits.  I believe that this is the last outstanding patch required to merge; I will approve and merge this PR after your update

Comment (gjc13) [reply src/device-manager/java/src/nl/Weave/DeviceManager/NetworkInfo.java]: Updated

**Pull request merged**